opta/upgrade.py

```diff
@@ -25,28 +27,37 @@ def _get_latest_version() -> str:
     return resp.text.strip().strip("v")
 
 
-def check_version_upgrade() -> None:
+def check_version_upgrade(is_upgrade_call: bool = False) -> bool:
     """Logs a warning if newer version of opta is available.
 
     The version check is not always performed when this function is called.
     It is performed non-deterministically with a probability of UPGRADE_CHECK_PROBABILITY
     in order to not spam the user.
     """
-    if not _should_check_for_version_upgrade():
-        return
-    logger.debug("Checking for version upgrades...")
-    try:
-        latest_version = _get_latest_version()
-    except Exception as e:
-        logger.debug(e, exc_info=True)
-        logger.debug("Unable to find latest version.")
-        return
-    try:
-        if semver.VersionInfo.parse(VERSION.strip("v")).compare(latest_version) < 0:
-            logger.warning(
-                "New version available.\n"
-                f"You have {VERSION} installed. Latest version is {latest_version}.\n"
-                f"Upgrade instructions are available at {UPGRADE_INSTRUCTIONS_URL}."
-            )
-    except Exception as e:
-        logger.debug(f"Semver check failed with error {e}")
+    if Path(successfull_upgrade).is_file():
+        os.remove(successfull_upgrade)
+        return False
+    if is_upgrade_call or _should_check_for_version_upgrade():
+        logger.info("Checking for version upgrades...")
```

Comment:
surprisingly when I test on main branch, this message was displayed twice sometimes:
```
(.venv) ➜  opta git:(feat/nilesh/runx-1037-opta-upgrade) ✗ python3 $PYTHONPATH/opta/cli.py upgrade                               
Checking for version upgrades...
Checking for version upgrades...
(.venv) ➜  opta git:(feat/nilesh/runx-1037-opta-upgrade) ✗ python3 $PYTHONPATH/opta/cli.py upgrade --help   
Usage: cli.py upgrade [OPTIONS]

  Upgrade Opta to the Latest version available

Options:
  -h, --help  Show this message and exit.
(.venv) ➜  opta git:(feat/nilesh/runx-1037-opta-upgrade) ✗ python3 $PYTHONPATH/opta/cli.py upgrade       
Checking for version upgrades...
(.venv) ➜  opta git:(feat/nilesh/runx-1037-opta-upgrade) ✗ python3 $PYTHONPATH/opta/cli.py upgrade 
Checking for version upgrades...
(.venv) ➜  opta git:(feat/nilesh/runx-1037-opta-upgrade) ✗ python3 $PYTHONPATH/opta/cli.py upgrade 
Checking for version upgrades...
Checking for version upgrades...
(.venv) ➜  opta git:(feat/nilesh/runx-1037-opta-upgrade) ✗ python3 $PYTHONPATH/opta/cli.py upgrade 
Checking for version upgrades...
(.venv) ➜  opta git:(feat/nilesh/runx-1037-opta-upgrade) ✗ python3 $PYTHONPATH/opta/cli.py upgrade 
Checking for version upgrades...
(.venv) ➜  opta git:(feat/nilesh/runx-1037-opta-upgrade) ✗ python3 $PYTHONPATH/opta/cli.py upgrade 
Checking for version upgrades...
(.venv) ➜  opta git:(feat/nilesh/runx-1037-opta-upgrade) ✗ python3 $PYTHONPATH/opta/cli.py upgrade 
Checking for version upgrades...

```

with debug flag
```
python3 $PYTHONPATH/opta/cli.py upgrade 
INFO: Checking for version upgrades...
DEBUG: Querying https://dev-runx-opta-binaries.s3.amazonaws.com/latest
DEBUG: Semver check failed with error  is not valid SemVer string
INFO: Checking for version upgrades...
DEBUG: Querying https://dev-runx-opta-binaries.s3.amazonaws.com/latest
DEBUG: Semver check failed with error  is not valid SemVer string
```

probably this call https://github.com/run-x/opta/blob/main/opta/cli.py#L86

Reply:
This happens only when the Version.txt file doesn't contain the correct Version. In that case, the upgrade itself won't happen. But we don't need to worry about that as of now, because we are maintaining the versions manually.

If you wish to test `opta upgrade`, please add correct version to the version.txt file. You will find the expected results.
And the reason you are sometimes getting the `Checking for version upgrades` because version check only happens in 20% of the runs.

Reply:
thank you for the explanation, I didn't know about this, where is the 20% rule defined in the code?

Reply:
Here you go: [LOC](https://github.com/run-x/opta/blob/main/opta/upgrade.py#L15)